### ISSUE TITLE
Automatically add . ConfigureAwait(false) when extracting an async method

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#38529]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/content/idea/705733/automatically-add-configureawaitfalse-when-extract.html)._

---
When using the c# Extract Method feature, if the method created is an async method, Visual Studio should automatically add . ConfigureAwait(false) to the end of the call to the new method it replaces the original code with, if the final await statement in the extracted code (on any branch of execution) has . ConfigureAwait(false)

---
#### Jane Wu [MSFT] on 8/23/2019, 00:08 AM: 

<p>Thank you for taking the time to provide your suggestion. We will do some preliminary checks to make sure we can proceed further.  We’ll provide an update once the issue has been triaged by the product team.</p>

**Comment by CyrusNajmabadi** [comment]
That doesn't seem correct to me.  Indeed, it would be *incorrect* in many cases, especially for the common case of any sort of UI app, or anything using JTF, etc. 

It might be resonable to analyze the surrounding code to make a reasonable guess about the appropriate .ConfigureAwait behavior.  But defaulting to `.ConfigureAwait(false)` seems actually undesirable.

**Comment by sharwell** [comment]
@CyrusNajmabadi The key for correctness here is this:

> if the final await statement in the extracted code (on any branch of execution) has . ConfigureAwait(false)

The code originally has `.ConfigureAwait(false)` before extraction, then it needs to retain it after extraction for correctness. For example:

```csharp
await Task MyDelay(TimeSpan duration) {
  await Task.Delay(duration).ConfigureAwait(false); // Select this line and use Extract Method
}
```

**Comment by YairHalberstadt** [comment]
I'm working on this at the moment.

My logic is as follows. If any of the selection contains `configureAwait(false)`, then the execution would not have continued on the captured context for the rest of the method after that point.

Therefore we should add `.ConfigureAwait(false)` to the new created invocation as this will maintain that behaviour.

This is the case even if there is a mixture of `ConfigureAwait(true)` and `false`.

Note that i cannot actually think of any use case for ConfigureAwait(false) followed by ConfigureAwait(true).

However this will have the correct behaviour when we have `ConfigureAwait(true)` followed by `ConfigureAwait(false)`. This can be useful if you update the UI in the method after the first await, but no longer do so for the rest of the method.

I only look for literal false. Any more complex expression is ignored.

**Comment by allisonchou** [comment]
Fixed by #38582
